apps/questdk/src/tokens/nft.ts

```diff
@@ -0,0 +1,60 @@
+import { type Address, type PublicClient, zeroAddress } from 'viem'
+import { fetchERC1155Metadata } from './erc1155'
+import { fetchERC721Metadata } from './erc721'
+import fetchQuestActionParams from '../quests/fetchQuestData.js'
+
+export async function fetchTokenMetadata(
+  client: PublicClient,
+  contractAddress: string,
+  tokenId: number,
+): Promise<any> {
+  try {
+    // Try to call the ERC721 balanceOf function
+    await client.readContract({
+      address: contractAddress as Address,
+      abi: [
+        'function balanceOf(address owner) external view returns (uint256)',
+      ],
+      functionName: 'balanceOf',
+      args: [zeroAddress],
+    })
+
+    // If the call succeeds, it's an ERC721 contract
+    return fetchERC721Metadata(client, contractAddress, tokenId)
+  } catch (error) {
+    // If the call fails, try to call the ERC1155 balanceOf function
+    await client.readContract({
+      address: contractAddress as Address,
+      abi: [
+        'function balanceOf(address account, uint256 id) external view returns (uint256)',
```

Comment:
what's funny is that this could still fail 🙃 I've seen reject when tokenId doesn't exist

Reply:
Yeah, it's not perfect. If the `tokenId` doesn't exist though we can't return metadata for it so 🤷